sql/opt: Generate synthetic check constraint to enforce RLS policies for new rows

spilchen opened the pull request:
With row-level security, policies include a WITH CHECK expression to enforce constraints on new rows. This commit begins adding support for enforcing these policies by modifying the optbuilder to construct the check constraint, evaluate the expression, and pass the result to the execution engine. A future commit will integrate the execution engine to fully enforce these policies.

Since the expression for the synthetic check constraint is determined at INSERT or UPDATE time, a placeholder check constraint is added when building the optimizer table catalog. The check constraint is then finalized in the mutationBuilder.

Because the check constraint is constructed late in the process, a function is needed to look up the column ordinal for a given column ID. To facilitate this, the previously internal function lookupColumnOrdinal has been made external as LookupColumnOrdinal.

Epic: CRDB-45203
Release note: None
Informs: #136704

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/141614)
<!-- Reviewable:end -->

DrewKimball reviewed (commented):
> Reviewed 3 of 11 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/141614)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @mgartner and @michae2)

___
*[`pkg/sql/opt_catalog.go` line 1146 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/141614#-OJeqEA33AerB7n20GJ0:-OJeqEA33AerB7n20GJ1:b-zi73ge) ([raw file](https://github.com/cockroachdb/cockroach/blob/ac6505a35e1090cc97174965fa901ac9562171f4/pkg/sql/opt_catalog.go#L1146)):*
> ```Go
> 		synthesizedChecks = append(synthesizedChecks, optCheckConstraint{
> 			isRLSConstraint: true,
> 		})
> ```

It seems unfortunate to introduce a new type of check constraint to the list that has to be handled differently and requires new plumbing. Would it be possible to make `mutationBuilder.addCheckConstraintCols` directly check for RLS policies rather than adding a dummy constraint?
___
*[`pkg/sql/opt/optbuilder/mutation_builder.go` line 894 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/141614#-OJeoxTr16lJO1NNxr_p:-OJeoxTr16lJO1NNxr_q:b-bqmyjv) ([raw file](https://github.com/cockroachdb/cockroach/blob/ac6505a35e1090cc97174965fa901ac9562171f4/pkg/sql/opt/optbuilder/mutation_builder.go#L894)):*
> ```Go
> 					tabMeta:  mb.md.TableMeta(mb.tabID),
> 					oc:       mb.b.catalog,
> 					user:     mb.checkPrivilegeUser,
> ```

nit: it looks like this is the only place where `mutationBuilder.checkPrivilegeUser` is used. Can we get rid of it and just do `mb.b.checkPrivilegeUser` here instead?


<!-- Sent from Reviewable.io -->

spilchen reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/141614)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @DrewKimball, @mgartner, and @michae2)

___
*[`pkg/sql/opt_catalog.go` line 1146 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/141614#-OJeqEA33AerB7n20GJ0:-OJrtgBYAJAMPFgxj2Y7:b-tq1xcx) ([raw file](https://github.com/cockroachdb/cockroach/blob/ac6505a35e1090cc97174965fa901ac9562171f4/pkg/sql/opt_catalog.go#L1146)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

It seems unfortunate to introduce a new type of check constraint to the list that has to be handled differently and requires new plumbing. Would it be possible to make `mutationBuilder.addCheckConstraintCols` directly check for RLS policies rather than adding a dummy constraint?
</blockquote></details>

Yes, it’s possible. In an earlier revision, I considered this approach. However, I ultimately decided to include a dummy check constraint in the `Table` object for several reasons. Handling it differently would have required adding special cases in multiple places:

-   In `mutationBuilder.init`, the logic for computing slice sizes had to be adjusted by adding +1 for checks if the table had RLS enabled.
-   Refactoring `mutationBuilder.addCheckConstraintCols`. This function needed modifications to properly handle the RLS constraint. Since we only perform actions when `mb.tab.CheckCount() != 0`, the logic had to be expanded to include RLS tables. Additionally, the per-constraint logic (the inner `for` loop) had to be modularized into a separate function/closure so it could be executed one additional time for RLS tables.
-   Correcting the `mutationBuilder.checkColIDs` comment. The existing comment states, _"Its length is always equal to the number of check constraints on the table (see opt.Table.CheckCount)."_ While changing this comment wouldn’t be a major issue, I felt that the original description was simple and intuitive. Keeping it unchanged avoided unnecessary complexity.

Since we were already treating it as a check constraint in multiple places, it made sense to explicitly add a constraint for it when constructing the `Table` object. Plus, we already had precedence for adding synthetic check constraints in the `Table` object for UDTs, so incorporating one for RLS felt like a natural fit.


<!-- Sent from Reviewable.io -->

michae2 reviewed (commented):
Nice! Just a couple questions.

> Reviewed 10 of 11 files at r1, 1 of 1 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/141614#-:-OJzBdP15UHfV9yIUKT9:b-er1z81)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @DrewKimball and @mgartner)

___
*[`pkg/sql/opt/testutils/testcat/test_catalog.go` line 1075 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/141614#-OJz2qDY1IddU1PlpTma:-OJz2qDY1IddU1PlpTmb:b-fv6uah) ([raw file](https://github.com/cockroachdb/cockroach/blob/5136c0c6a78bfccd8141e39e640c51a4e7d5281b/pkg/sql/opt/testutils/testcat/test_catalog.go#L1075)):*
> ```Go
> 		}
> 	}
> 	return 0, nil
> ```

Should this be an error?
___
*[`pkg/sql/opt/optbuilder/row_level_security.go` line 132 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/141614#-OJz2ANJEQupvSTdS8Me:-OJz2ANJEQupvSTdS8Mf:b-g3dww2) ([raw file](https://github.com/cockroachdb/cockroach/blob/5136c0c6a78bfccd8141e39e640c51a4e7d5281b/pkg/sql/opt/optbuilder/row_level_security.go#L132)):*
> ```Go
> 		constraint:  expr,
> 		columnCount: len(colIDs),
> 		lookupColumnOrdinal: func(i int) (int, error) {
> ```

nit: Would it be simpler to save `r.tab` and `colIDs` in the struct and do the call to `LookupColumnOrdinal` directly in `ColumnOrdinal`?
___
*[`pkg/sql/opt/optbuilder/testdata/row_level_security` line 546 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/141614#-OJz5il96jnVzRg_a_0r:-OJz5il96jnVzRg_a_0s:b-y1sspx) ([raw file](https://github.com/cockroachdb/cockroach/blob/5136c0c6a78bfccd8141e39e640c51a4e7d5281b/pkg/sql/opt/optbuilder/testdata/row_level_security#L546)):*
> ```
> 
> build
> INSERT INTO t1_explicit_pk VALUES (2, 'second', '2008-11-19') ON CONFLICT DO NOTHING;
> ```

Should SELECT policies apply to the read part of INSERT ON CONFLICT?
___
*[`pkg/sql/opt/optbuilder/mutation_builder.go` line 903 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/141614#-OJybBsE6YEa3CpkZyXE:-OJybBsE6YEa3CpkZyXF:bhvhbdm) ([raw file](https://github.com/cockroachdb/cockroach/blob/5136c0c6a78bfccd8141e39e640c51a4e7d5281b/pkg/sql/opt/optbuilder/mutation_builder.go#L903)):*
> ```Go
> 			colName := scopeColName("")
> 			if check.IsRLSConstraint() {
> 				colName = colName.WithMetadataName("rls")
> ```

Is there an assertion somewhere that IsRLSConstraint is true for only one of the check constraints?
___
*[`pkg/sql/opt/optbuilder/mutation_builder.go` line 923 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/141614#-OJzBNo97lygpnFJ8F3z:-OJyhiWZ29UPvC76TEyR:b-sq06p9) ([raw file](https://github.com/cockroachdb/cockroach/blob/5136c0c6a78bfccd8141e39e640c51a4e7d5281b/pkg/sql/opt/optbuilder/mutation_builder.go#L923)):*
> ```Go
> 			//   expressions can exist for read and write operations. This means it's
> 			//   possible to read a row whose column values would violate the write
> 			//   expression.
> ```

Is there a test case that validates this?


<!-- Sent from Reviewable.io -->

spilchen committed:
sql/opt: Generate synthetic check constraint to enforce RLS policies for new rows

With row-level security, policies include a WITH CHECK expression to
enforce constraints on new rows. This commit begins adding support for
enforcing these policies by modifying the optbuilder to construct the
check constraint, evaluate the expression, and pass the result to the
execution engine. A future commit will integrate the execution engine to
fully enforce these policies.

Since the expression for the synthetic check constraint is determined at
INSERT or UPDATE time, a placeholder check constraint is added when
building the optimizer table catalog. The check constraint is then
finalized in the mutationBuilder.

Because the check constraint is constructed late in the process, a
function is needed to look up the column ordinal for a given column ID.
To facilitate this, the previously internal function lookupColumnOrdinal
has been made external as LookupColumnOrdinal.

Epic: CRDB-45203
Release note: None
Informs: #136704

spilchen reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/141614)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @DrewKimball, @mgartner, and @michae2)

___
*[`pkg/sql/opt/optbuilder/mutation_builder.go` line 903 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/141614#-OJybBsE6YEa3CpkZyXE:-OK30D3R67c5wfNLiAUD:bwsm5tf) ([raw file](https://github.com/cockroachdb/cockroach/blob/5136c0c6a78bfccd8141e39e640c51a4e7d5281b/pkg/sql/opt/optbuilder/mutation_builder.go#L903)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

Is there an assertion somewhere that IsRLSConstraint is true for only one of the check constraints?
</blockquote></details>

Not yet, but that's a good suggestion. I'll add one to mutationBuilder.addCheckConstraintCols when we walk all of the constraints.
___
*[`pkg/sql/opt/optbuilder/mutation_builder.go` line 923 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/141614#-OJzBNo97lygpnFJ8F3z:-OK3-C7Z701mZVpFieAW:belfrv0) ([raw file](https://github.com/cockroachdb/cockroach/blob/5136c0c6a78bfccd8141e39e640c51a4e7d5281b/pkg/sql/opt/optbuilder/mutation_builder.go#L923)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

Is there a test case that validates this?
</blockquote></details>

There is a test for this in opt/optbuilder/testdata/row_level_security. Search for this comment:
```
# Verify that an update fetches the column if a policy references it, even when
# the column itself is not being modified.
```

The test verifies it by including the columns fetched in the plan. When I do the execution engine side of the changes, I was going to have a logictest for this too.
___
*[`pkg/sql/opt/testutils/testcat/test_catalog.go` line 1075 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/141614#-OJz2qDY1IddU1PlpTma:-OK2zilE8iKlUNSJCL3-:b-f06h8t) ([raw file](https://github.com/cockroachdb/cockroach/blob/5136c0c6a78bfccd8141e39e640c51a4e7d5281b/pkg/sql/opt/testutils/testcat/test_catalog.go#L1075)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

Should this be an error?
</blockquote></details>

Yes, good catch!
___
*[`pkg/sql/opt/optbuilder/testdata/row_level_security` line 546 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/141614#-OJz5il96jnVzRg_a_0r:-OK30Uj6CXA0GULqy_tL:byacrmw) ([raw file](https://github.com/cockroachdb/cockroach/blob/5136c0c6a78bfccd8141e39e640c51a4e7d5281b/pkg/sql/opt/optbuilder/testdata/row_level_security#L546)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

Should SELECT policies apply to the read part of INSERT ON CONFLICT?
</blockquote></details>

Yes! And we need to apply the INSERT and UPDATE policies if we are going to follow the postgres model. I created this test to understand the behaviour when I realized that there is a bunch more work needed to support upsert. I ended up creating an issue (https://github.com/cockroachdb/cockroach/issues/141998) to deal with upsert rather than add it here. I'll add a comment to this testcase to indicate the support for it isn't complete.


<!-- Sent from Reviewable.io -->

DrewKimball reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 4 of 11 files at r1, 4 of 4 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/141614#-:-OKSGFY06kGjbfaUUmpk:bnfp4nl)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @mgartner)



<!-- Sent from Reviewable.io -->

michae2 reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> Nice!

> Reviewed 4 of 4 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/141614#-:-OKT8Hik5seo7tpBv4fd:b-8gk8uj)* status: <span>:shipit:</span> complete! 2 of 0 LGTMs obtained (waiting on @mgartner)

___
*[`pkg/sql/opt/optbuilder/mutation_builder.go` line 923 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/141614#-OJzBNo97lygpnFJ8F3z:-OKT8kpP9N38-jxpvu-e:b-nff86g) ([raw file](https://github.com/cockroachdb/cockroach/blob/5136c0c6a78bfccd8141e39e640c51a4e7d5281b/pkg/sql/opt/optbuilder/mutation_builder.go#L923)):*
<details><summary><i>Previously, spilchen wrote…</i></summary><blockquote>

There is a test for this in opt/optbuilder/testdata/row_level_security. Search for this comment:
```
# Verify that an update fetches the column if a policy references it, even when
# the column itself is not being modified.
```

The test verifies it by including the columns fetched in the plan. When I do the execution engine side of the changes, I was going to have a logictest for this too.
</blockquote></details>

Ah, thank you!
___
*[`pkg/sql/opt/optbuilder/testdata/row_level_security` line 546 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/141614#-OJz5il96jnVzRg_a_0r:-OKT9L6A1kR6ehgD_7Tj:b-xtgcr2) ([raw file](https://github.com/cockroachdb/cockroach/blob/5136c0c6a78bfccd8141e39e640c51a4e7d5281b/pkg/sql/opt/optbuilder/testdata/row_level_security#L546)):*
<details><summary><i>Previously, spilchen wrote…</i></summary><blockquote>

Yes! And we need to apply the INSERT and UPDATE policies if we are going to follow the postgres model. I created this test to understand the behaviour when I realized that there is a bunch more work needed to support upsert. I ended up creating an issue (https://github.com/cockroachdb/cockroach/issues/141998) to deal with upsert rather than add it here. I'll add a comment to this testcase to indicate the support for it isn't complete.
</blockquote></details>

Thanks, that makes sense.


<!-- Sent from Reviewable.io -->
